### DISCUSSION TITLE
Beginner's question about handling of effects

### DISCUSSION THREAD

**Discussion by AZMCode** [koka-lang/koka#366] [Q&A] [answered]
I'm currently in the process of defining an effect for logging, and created a small handler to simply print the logged strings to console. However, I'm getting an "effects do not match" error when I'm pretty sure the code is well typed and well effected. Here's the code in question:

```
pub linear effect fun log(s: string): ()

pub fun log-println<e>(body: () -> <log|e> a): <console|e> a
  with fun log(s) println(s) 
  body()
```

and the error is
```
log.kk(9, 3): error: effects do not match
  context        :   with fun log(s) println(s) 
                     body()
  term           :   body()
  inferred effect: <log|$e>
  expected effect: <log,console|_e1>
```

Is this to do with the inner body of the handler using the console effect? Do I need to defer logging to the outside expression? Just confused overall.

**Comment by TimWhiting** [comment, marked as answer]
Effects are all about dynamic scope. 

Since console is required to be in the dynamic scope for this function, any function you pass in will have access to console in addition to log & the polymorphic effect `e`. The problem is that `e` could unify with another `console` effect, so we need to be clear in the type that body has access to console from this same scope. This way, if body introduces it's own console effect, it's type will include two console effects, and the usages within the body will interact with it's own console effect, and only usages outside it's installed handler or usages that mask the innermost handler will select the one in your function's context. 

This doesn't make as much sense for console, since it is a builtin effect, and cannot be overridden or have another handler installed, but the type system doesn't distinguish between builtin effects and user handled effects in this case.

**Reply by chtenb** [reply]
Thanks. This area could do with more elaborate documentation (or error messages) I think, because when starting with Koka I ran into this a lot as well without being able to understand why.

**Reply by AZMCode** [reply]
Thanks for the clarification. I can't quite qrap my head about this right now, but I'll have some time to think about it later. I appreciate the answer though!
